core/domain/exp_services_test.py

```diff
@@ -4052,10 +4052,45 @@ def test_update_interaction_solutions(self):
             })], 'Changed interaction_solutions.')
 
         exploration = exp_fetchers.get_exploration_by_id(self.NEW_EXP_ID)
+
         self.assertEqual(
             exploration.init_state.interaction.solution.to_dict(),
             solution)
 
+    def test_update_solution_no_interaction_id(self):
```

Comment:
Thanks for adding this test, can you add one more test for `invalid interaction id` i.e, the expected interaction id was X but passed Y.

Reply:
Sorry for the delays.

The test that I added fails with _key_error_ inside the _exp_services_ code, while trying to create the **Solution** object with _from_dict_.

It is not testing that the "update" has the wrong interaction_id (it can't find it), but rather that it does not have it set at all. 
Maybe I could find a better name for the test (_forgot_interaction_id_update_ ??).

I tried to do what you are asking but I think that the **"expected X but passed Y"** case is not really reachable from this part of the code. 
The interaction ids are also refreshed at each _Registry.get_interaction_by_id_ call

The code always takes the interaction_id of the state (based on state name). We already test the _"wrong state name passed"_ case (it is a common step for all types of updates).

Reply:
> The test that I added fails with key_error inside the exp_services code,

Can you show me the stack trace? (I wanted to see the line of code where it fails?)

If it's not possible to test this through the service layer, can we add this test to the domain layer of State object, which will test the update_interaction_sultion function? (maybe)

Reply:
This is the stacktrace:
```
Traceback (most recent call last):
  File "/oppia/core/domain/exp_services_test.py", line 4090, in test_update_solution_no_interaction_id
    'Changed interaction_solutions.'
  File "/oppia/core/domain/exp_services.py", line 878, in update_exploration
    updated_exploration = apply_change_list(exploration_id, change_list)
  File "/oppia/core/domain/exp_services.py", line 384, in apply_change_list
    state.interaction.id, change.new_value
  File "/oppia/core/domain/state_domain.py", line 249, in from_dict
    interaction_id).normalize_answer(
  File "/oppia/core/domain/interaction_registry.py", line 86, in get_interaction_by_id
    return cls._interactions[interaction_id]
KeyError: None
```

As for the _update_interaction_solution_ method:
The method now already receives a **Solution** object.  The object must be created from outside and this check is done only at the time you create the object.

The _interaction_id_ parameter forces us to create objects from outside (where you have access to it). As long as the state of the change is correct, the update will go through (otherwise it will fail sooner).

correct state  == exists and it has an interaction id 
I added a test for the second condition and there are tests for the first (the update fails faster in that case).

Reply:
Thanks for the explanation, I think we are having some miscommunication here, I wanted to check this:

Assume a state has interaction_id TextInput (valid intractionId), let's create a Solution object using MultipleChoiceInput (valid interaction) and pass this solution to state.update_interaction_solution and check whether this update fails? Let me know if it's still not clear?

Reply:
Sorry for the misunderstanding.
I had to do some debugging to understand what the interaction_id actually means, the naming is quite confusing (as it is with the way they are populated, inside the registry). It's more like an interaction type (--> I thought that there are many interactions of the same type and the different **IDs** and then there's some equals check somewhere for each interaction).

I created the test case that you are talking about and it does not fail, even though the solution has now a new type. It just replaces the solution to point to our new object.

Reply:
Thanks for taking a deeper look @FlorinBalint and sharing your findings! :)

[Wanted to make sure that you have validate the exploration after saving the new (buggy) solution?] 

So I think we need validation to make sure that the solution is created with a give interaction is? Can we do something for that? Also, can you think of a case where this can be an issue i.e, a solution created via a different interactionID? (If there isn't an issue then we can leave it as - is. @seanlip, do you have any suggestions?)

Reply:
@FlorinBalint, I had a discussion with sean for the same and I think we should add validation in the state object to make sure "if normalize(solution.interaction.id, solution.answer) != solution.correct_answer", raise an error. Can you please whether we can implement this?

Let me know if you have any questions? Thanks! :)